src/Compiler/Driver/ReuseTcResults/TcResultsPickle.fs

```diff
@@ -0,0 +1,1745 @@
+
+let pickleTcState (tcState: TcState) (st: WriterState) =
+    p_ccuref_new tcState.tcsCcu st
+    p_tc_env tcState.tcsTcSigEnv st
+    p_tc_env tcState.tcsTcImplEnv st
+    p_bool tcState.tcsCreatesGeneratedProvidedTypes st
+    (p_list p_tcs_root_sig) (tcState.tcsRootSigs.ToList()) st
+    p_list p_qualified_name_of_file (tcState.tcsRootImpls.ToList()) st
+    p_modul_typ_new tcState.tcsCcuSig st
+    p_list p_open_decl tcState.tcsImplicitOpenDeclarations st
+
+// unpickling
+
+let u_stamp = u_int64
+
+let u_stamp_map uv = u_Map u_stamp uv 
+
+let u_non_null_slot f st =
+    let tag = u_byte st
+    match tag with
+    | 0 -> Unchecked.defaultof<_>
+    | 1 -> f st
+    | n -> ufailwith st ("u_option: found number " + string n)
+
+
+let u_ILTypeDefAdditionalFlags st : ILTypeDefAdditionalFlags =
+    let i = u_int32 st
+    enum i
+
+let u_ILTypeDef st : ILTypeDef = 
+    let name = u_string st
+    let attributes = System.Reflection.TypeAttributes.Public
+    let layout = ILTypeDefLayout.Auto
+    let implements = Unchecked.defaultof<_>
+    let genericParams = []
+    let extends = Unchecked.defaultof<_>
+    let methods = ILMethodDefs(fun () -> [||])
+    let nestedTypes = Unchecked.defaultof<_>
+    let fields = Unchecked.defaultof<_>
+    let methodImpls = Unchecked.defaultof<_>
+    let events = Unchecked.defaultof<_>
+    let properties = Unchecked.defaultof<_>
+    let additionalFlags = u_ILTypeDefAdditionalFlags st
+    let securityDeclsStored = ILSecurityDecls([||])
+    // TODO: fill this in
+    let customAttrsStored = ILAttributesStored.Given (ILAttributes.Empty)
+
+    ILTypeDef(name,
+        attributes,
+        layout,
+        implements,
+        genericParams,
+        extends,
+        methods,
+        nestedTypes,
+        fields,
+        methodImpls,
+        events,
+        properties,
+        additionalFlags,
+        securityDeclsStored,
+        customAttrsStored)
+
+let u_tyar_spec_data_new st =
+    let a, c, d, e, g, stamp = u_tup6 u_ident u_attribs u_int64 u_tyar_constraints u_xmldoc u_stamp st
+    { typar_id=a
+      typar_stamp=stamp
+      typar_flags=TyparFlags(int32 d)
+      typar_solution=None
+      typar_astype= Unchecked.defaultof<_>
+      typar_opt_data=
+        match g, e, c with
+        | doc, [], [] when doc.IsEmpty -> None
+        | _ -> Some { typar_il_name = None; typar_xmldoc = g; typar_constraints = e; typar_attribs = c;typar_is_contravariant = false } }
+
+let u_tyar_spec_new st =
+    u_osgn_decl st.itypars u_tyar_spec_data_new st
+
+let u_tyar_specs_new = u_list u_tyar_spec_new
+
+let rec u_ty_new st : TType =
+    let tag = u_byte st
+    match tag with
+    | 0 ->
+        let tupInfo, l =
+            u_tup2
+                u_tup_info
+                u_tys_new 
+                st
+        TType_tuple (tupInfo, l)
+
+    | 1 ->
+        let tyconRef, typeInstantiation, nullness, binding =
+            u_tup4
+                u_tcref
+                u_tys_new
```

Comment:
That's really very, very hard. I haven't come up with a mechanism to properly cache and restore tcref. The compiler needs different info about the entity on different stages (optimizations, tail call checks, code gen, whatever...). And goes through things like typar solutions and referenced entities. But storing it all fully leads to StackOverflows usually...